text/0000-all-builder-buildpackless.md

```diff
@@ -0,0 +1,76 @@
+# Make All Builders Buildpackless
+
+## Summary
+
+Builders that conglomerate buildpacks are unsustainable in the long run as
+there is a layer cap set in place by Docker of 127 layers. We should stop
+adding buildpack to our builder and instead only offer builders that contain
+the stack and lifecycle. 
+
+## Motivation
+
+With the current size of the `full` builder we are already seeing users running
+into issues with exceeding the layer limit of the buildpack when trying to add
+custom buildpacks to their builds. This will only be exacerbated going forward
+with the addition of new buildpacks into the Paketo project. Not only that the
+`full` builders are completely unsustainable in the near future with proposals
+such as the addition of APMs and the restructuring of existing buildpacks which
+will both increase the layer count further.
+
+Currently the purpose of the builders that include buildpacks is to allow first
+time users to quickly plug in a large number of applications into the build
+process and see what comes out. However using these loaded builders is not the
+optimal solution if you are running builds for production images and the
+builder images are large and bloated for most users who only have applications
+in a couple of language domains.
+
+This layer cap will only continue to become more of a problem and there is no
+clear workaround so unfortunately a drastic restructuring is required.
+
+## Rationale and Alternatives
+
+### Language Specific Builders
+Have one builder per language family that only supports that language family.
+
+#### Pro
+This guarantees that the builder will remain small as they will be confined to
+a much smaller domain.
+
+#### Cons
+There is not much difference not much between this workflow and specifying the
+buildpack that you want during build with a builder that contains no
+buildpacks.
```

Comment:
I think the real "con" here is that we would have to maintain ~10x more builders than we currently do, and as you point out, the effort of finding a particular language builder for an app is about the same level of effort as running a buildpackless builder with a `--buildpack` flag.

Reply:
From a user stand point, using `-b` is slower. It dynamically creates a new builder every time you do a `pack build`. If you have small buildpacks, like our Paketo Buildpacks, it generally goes quick. If you embed dependencies in your buildpack images, it'll be very, very slow.

Reply:
>I think the real "con" here is that we would have to maintain ~10x more builders than we currently do

This sounds like a process scalability problem. Perhaps it is an opportunity to rearchitect how we manage and ship builders in a more scalable way? Certainly not a short-term solution, but perhaps something to consider.

Reply:
I see the performance implications of always using `-b` as a pretty sizeable con. We have a lot of evidence that performance is a major concern to buildpack users and one of the first things potential users evaluate when comparing tools.

Reply:
> I think the real "con" here is that we would have to maintain ~10x more builders than we currently do

I agree with Dan that this sounds like a process/automation problem? Whether we publish the language-specific builders or only buildpackless builders we still need to test and maintain the same combinatorial matrix of buildpack+stack+lifecycle. Ideally, publish more images to improve performance for end users shouldn't materially increase increase that burden.